### PR TITLE
[13.0][IMP]storage_image_product: main image of the variant

### PR DESCRIPTION
**Pull request by AaronHForgeFlow**
@ForgeFlow @HviorForgeFlow

### CODE REVIEW

**Comment by simahawk** [comment]
this is similar to https://github.com/OCA/storage/pull/116 but for images right?

**Comment by AaronHForgeFlow** [comment]
When the images are attached to the template and not to the variant. In that case in the variant it shows an image without attributes. The fix search first for images with the same attributes than the variant. 

Perhaps this is not a common use case as long I am storing all the images in the template instead of storing them in any image individually. If that is the case I will close this.

**Comment by sebastienbeau** [comment]
Hi
I think I understand.
If you put 3 images on a product template in this order
- a generic image for all product (like a image of the situtation, or an image with all the variant visible)
- a specific image for the blue product
- a specific image for the red product

With the existing implementation, the generic one will used as main image for the product template and for all variant.
With your change each variant will try to show the first "specific image" that match for them. So in our case it will a be the blue image for the blue product, the red image for the red product.

I am correct ?

If yes it's seem to be a good idea, can you add a test? If needed i can do it

Thanks

**Comment by AaronHForgeFlow** [comment]
> If you put 3 images on a product template in this order
> 
> a generic image for all product (like a image of the situtation, or an image with all the variant visible)
> a specific image for the blue product
> a specific image for the red product
> With the existing implementation, the generic one will used as main image for the product template and for all variant.
> With your change each variant will try to show the first "specific image" that match for them. So in our case it will a be the blue image for the blue product, the red image for the red product.

Yes, exactly :)

I will try to add a test Thanks for the review!

**Comment by AaronHForgeFlow** [comment]
Test added :)

File: storage_image_product/tests/test_product_image_relation.py [hunk 1]
```diff
@@ -114,3 +114,39 @@ def test_main_image_and_urls(self):
             (self.logo_image, self.product_b),
         )
         self._test_main_images_and_urls(expected)
+
+    def test_main_image_attribute(self):
+        """
+        Attach the image to the template and check the first image of the
+        variant is the one with same attriubttes
```
Comment (gurneyalex) [inline storage_image_product/tests/test_product_image_relation.py]:
```suggestion
        variant is the one with same attributes
```

File: storage_image_product/tests/test_product_image_relation.py [hunk 2]
```diff
@@ -114,3 +114,39 @@ def test_main_image_and_urls(self):
...
+                        [
+                            self.env.ref("product.product_attribute_value_4").id,
+                            self.env.ref("product.product_attribute_value_1").id,
+                        ],
+                    )
+                ],
+                "sequence": 10,
+            }
+        )
+        # The variant should not take the only wiht lower sequencce but
```
Comment (gurneyalex) [inline storage_image_product/tests/test_product_image_relation.py]:
```suggestion
        # The variant should not take the only with the lowest sequence but
```

**Review by gurneyalex — APPROVED** [review]

**Review by simahawk — APPROVED** [review]

**Comment by sebastienbeau** [comment]
@AaronHForgeFlow can you add the change propose by alex ? so we can merge this ? Thansk

**Comment by AaronHForgeFlow** [comment]
> @AaronHForgeFlow can you add the change propose by alex ? so we can merge this ? Thansk

Done! Thanks! :)

**Comment by sebastienbeau** [comment]
Thanks @AaronHForgeFlow for your work !

**Pull request merged**